src/workerd/api/memory-cache.h

```diff
@@ -0,0 +1,497 @@
+  };
+
+private:
+
+  // To ensure thread-safety, all mutable data is guarded by a mutex. Each cache
+  // operation requires an exclusive lock. Even read-only operations need to
+  // update the liveliness of cache entries, which currently requires a lock.
+  kj::MutexGuarded<ThreadUnsafeData> data;
+
+  // The MemoryCacheProvider instance needs to be guaranteed to outlive the SharedMemoryCache
+  // instance. When the SharedMemoryCache is destroyed, it will remove itself from the provider.
+  // TODO(cleanup): Eventually, assuming/once the kj::Ptr<T> work progresses, it would be safer
+  // to replace this with a kj::Ptr<MemoryCacheProvider>
+  kj::Maybe<const MemoryCacheProvider&> provider;
+
+  // The MemoryCacheProvider owns the underlying kj::String for the id, since the provider has
+  // to outlive this SharedMemoryCache instance, so does the id. When this SharedMemoryCache
+  // is destroyed and unregisters itself, the id will also be freed.
+  kj::StringPtr id;
+
+  // Same as above, the MemoryCacheProvider owns the actual handler here. Since that is guaranteed
+  // to outlive this SharedMemoryCache instance, so is the handler.
+  kj::Maybe<AdditionalResizeMemoryLimitHandler&> additionalResizeMemoryLimitHandler;
+};
+
+// JavaScript class that allows accessing an in-memory cache.
+// Each instance of this class holds a SharedMemoryCache::Use object and
+// all calls from JavaScript are essentially forwarded to that object, which
+// manages interaction with the shared cache in a thread-safe manner.
+class MemoryCache: public jsg::Object {
+public:
+  MemoryCache(SharedMemoryCache::Use&& use): cacheUse(kj::mv(use)) {}
+
+  using FallbackFunction = jsg::Function<jsg::Promise<CacheValueProduceResult>(kj::String)>;
+
+  // Reads a value from the cache or invokes a fallback function to obtain the
+  // value, if a fallback function was given.
+  jsg::Promise<jsg::JsRef<jsg::JsValue>> read(jsg::Lock& js,
+      jsg::NonCoercible<kj::String> key,
+      jsg::Optional<FallbackFunction> optionalFallback);
+
+  JSG_RESOURCE_TYPE(MemoryCache) { JSG_METHOD(read); }
+
+private:
+  SharedMemoryCache::Use cacheUse;
+};
+
+// The MemoryCacheProvider provides the internal implementation of the MemoryCache mechanism.
+// It is responsible for owning the SharedMemoryCache instances and providing them to the
+// bindings as needed. The default implementation (created and returned by createDefault())
+// uses a simple in-memory map to store the SharedMemoryCache instances.
+//
+// This class is virtualized for a number of reasons, the most important of which is that
+// we may want to replace the default implementation in production with one that includes
+// more metrics and monitoring, or that uses a different cache implementation. Ultimately,
+// however, if we do end up having additional implementations, it will be the SharedMemoryCache
+// class that will need to be updated.
+class MemoryCacheProvider {
```

Comment:
I still don't feel that this abstraction is adding value. I think we should just have `MemoryCacheMap` as a concrete type. I don't see how you could have any other implementation of this interface anyway.